src/transformers/models/gemma2/modeling_gemma2.py

```diff
@@ -747,8 +747,24 @@ def forward(
             inputs_embeds = self.embed_tokens(input_ids)
 
         if cache_position is None:
+            if past_key_values is not None and not past_key_values.empty:
+                logger.warning(
+                    "You are calling the model with non-empty `past_key_values` but didn't pass `cache_position`. ",
+                    "This will results in incorrect logits. Please pass `cache_position` that indicates indicates "
+                    "input's positions in the sequence.",
+                )
             cache_position = torch.arange(0, inputs_embeds.shape[1], device=inputs_embeds.device)
 
+        # Probably a forward call with caching, so we set up cache for one call only
+        if use_cache and past_key_values is None:
+            past_key_values = DynamicCache()
```

Comment:
Is there a reason why we don't initialize a `HybridCache` instead? As you wrote, we enter unpredictable territory when we go beyond 4k tokens :)

Reply:
Oh, actually we could. I just wasn't sire how to initialize max cache length, either as current input length or sliding window length. Do you think it's okay to make a `max-cache-len=input-len` so that users won't find a super long kv filled with zeros at the end of model call?

Reply:
Perhaps something like `input_length` rounded up to the next power of 2? (e.g. 100 -> 128, 300 -> 512)

We should throw a warning too